### PR TITLE
Implement `Eq`, `PartialEq`, `Hash` for `dyn PhysicalExpr`

### PR DESCRIPTION
**Pull request by peter-toth**
## Which issue does this PR close?

Part of https://github.com/apache/datafusion/issues/12599.

## Rationale for this change

Common subexpression elimination will require `Arc<dyn PhysicalExpr>` tree nodes to implement `Eq`, `PartialEq`, `Hash`.

## What changes are included in this PR?

This PR:
- Refactors `PhysicalExpr` trait and moves `dyn_hash()` to a new `DynHash` trait.
- Adds blanket implementation of `DynHash` for types that implement `Hash`. This enables us to drop most of the previous `dyn_hash()` boilerplate implementations.
- Introduces `DynEq` trait with `dyn_eq()`.
- Adds blanket implementation of `DynEq` for types that implement `Eq`.

Please note that:
- This PR contains an API breaking change that `dyn_hash()` is moved from `PhysicalExpr` to separate `DynHash` trait.
- This PR contains manual implementation of `PartialEq` and `Hash` for some `dyn PhysicalExpr` types (e.g. `BinaryExpr`) where adding just `#[derive(PartialEq, Hash)]` would be straightforward at first glance. 
   This is due to a Rust bug: https://github.com/rust-lang/rust/issues/78808 that the derive macro would generate code with the following issue:
   ```
      error[E0507]: cannot move out of `other.left` which is behind a shared reference
     --> datafusion/physical-expr/src/expressions/binary.rs:53:5
      |
   51 | #[derive(Debug, Hash, Clone, Eq, PartialEq)]
      |                                  --------- in this derive macro expansion
   52 | pub struct BinaryExpr {
   53 |     left: Arc<dyn PhysicalExpr>,
      |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^ move occurs because `other.left` has type `Arc<dyn PhysicalExpr>`, which does not implement the `Copy` trait
   ```

## Are these changes tested?

Yes, with existing tests.

## Are there any user-facing changes?

No.

### CODE REVIEW

**Comment by peter-toth** [comment]
cc @alamb

**Comment by alamb** [comment]
On my list for review tomorrow

File: datafusion/physical-expr/src/expressions/binary.rs [hunk 1]
```diff
@@ -525,20 +519,6 @@ impl PhysicalExpr for BinaryExpr {
     }
 }
 
-impl PartialEq<dyn Any> for BinaryExpr {
```
Comment (alamb) [inline datafusion/physical-expr/src/expressions/binary.rs:528]: removing this is great

File: datafusion/physical-expr/src/expressions/binary.rs [hunk 2]
```diff
@@ -48,11 +47,11 @@ use kernels::{
 };
 
 /// Binary expression
-#[derive(Debug, Hash, Clone)]
-pub struct BinaryExpr {
-    left: Arc<dyn PhysicalExpr>,
+#[derive(Debug, Hash, Clone, Eq, PartialEq)]
+pub struct BinaryExpr<DynPhysicalExpr: ?Sized = dyn PhysicalExpr> {
```
Comment (alamb) [inline datafusion/physical-expr/src/expressions/binary.rs:51]: The removal of the boiler plate in this PR is great ❤️ 

I feel like this additional trait / workaround is non ideal as it makes understanding the `BinaryExpr` very unobvious (there is now a level of indirection that is irrelevant for most of users of `BinaryExpr`).

Comment (alamb) [reply datafusion/physical-expr/src/expressions/binary.rs:51]: I am playing around with it to see if I can avoid it somehow

**Review by alamb — COMMENTED** [review]
Thank you @peter-toth  -- this PR is a pretty nice improvement in its own right (even without the improved CSE). 

Let's see if we can avoid making `BinaryExpr` generic

File: datafusion/physical-expr/src/expressions/binary.rs (hunk 2 above)
Comment (alamb) [reply datafusion/physical-expr/src/expressions/binary.rs:51]: Here is one way that seems to work: https://github.com/peter-toth/datafusion/pull/5

More verbose, but I think it keeps the structs simpler to understand

Comment (peter-toth) [reply datafusion/physical-expr/src/expressions/binary.rs:51]: Sure, we don't need to keep the generic parameter workaround. I agree that those params look a bit weird.
https://github.com/peter-toth/datafusion/pull/5 looks good to me, if you extend it to other expressions I can merge it tomorrow and amend my PR description.

Comment (alamb) [reply datafusion/physical-expr/src/expressions/binary.rs]: I think I accidentally push the commits to your branch 😬 -- hopefully that is ok

Comment (peter-toth) [reply datafusion/physical-expr/src/expressions/binary.rs]: No problem. :grin: I've just updated the PR description.

File: datafusion/physical-expr-common/src/physical_expr.rs [hunk 3]
```diff
@@ -183,6 +151,39 @@ pub trait PhysicalExpr: Send + Sync + Display + Debug + PartialEq<dyn Any> {
     }
 }
 
+pub trait DynEq {
```
[the remark this answers is not in the document]
Comment (peter-toth) [reply datafusion/physical-expr-common/src/physical_expr.rs:156]: Added in https://github.com/apache/datafusion/pull/13005/commits/f1917fa08e916efdbdce659b8ee71b7bdb2df0ba.

**Review by alamb — APPROVED** [review]
Thank you @peter-toth 

I think this is an improvement in my mind, even outside the context of CSE improvements, as it permits `PhysicalExpr ` to implement the standard `PartialEq` rather than a more special `DynEq` / `any` pattern

Perhaps @akurmustafa / @mustafasrepo may have some additonal thoughts as I think they added the original dyn hash work

cc @berkaysynnada

File: datafusion/physical-expr/src/expressions/case.rs [hunk 4]
```diff
@@ -80,7 +79,7 @@ enum EvalMethod {
 ///     [WHEN ...]
 ///     [ELSE result]
 /// END
-#[derive(Debug, Hash)]
+#[derive(Debug, Hash, PartialEq, Eq)]
```
Comment (berkaysynnada) [inline datafusion/physical-expr/src/expressions/case.rs:82]: How can it permit auto derivation of `PartialEq` for `CaseExpr` 🤔 -- we need manual impl's for others

**Review by berkaysynnada — APPROVED** [review]
I believe this PR brings significant improvements to the API, enhancing both code readability and usability. The clarity of the trait has increased, making it more intuitive to work with. LGTM, Thank you @peter-toth.

File: datafusion/physical-expr/src/expressions/case.rs (hunk 4 above)
Comment (peter-toth) [reply datafusion/physical-expr/src/expressions/case.rs:82]: Yeah, this is probably the strangest part of the bug. If you have double wrappers around `dyn Trait` then there is no issue: https://github.com/rust-lang/rust/issues/78808#issuecomment-1664012270
`CaseExpr` doesn't have any `dyn Trait` fields with a single wrapper so the `derive` macro has no problem.

Comment (berkaysynnada) [reply datafusion/physical-expr/src/expressions/case.rs:82]: Should we create a ticket to remove those impl's once the bug is resolved?

Comment (peter-toth) [reply datafusion/physical-expr/src/expressions/case.rs:82]: How about https://github.com/apache/datafusion/issues/13196?

**Comment by alamb** [comment]
What I would like to propose for this PR is wait until we have made the next release: #12470  and then we can merge it after that

I am thinking of how we can spread out the API upgrade pain

**Comment by peter-toth** [comment]
Thanks for review @alamb and @berkaysynnada.

**Pull request merged**